### ISSUE TITLE
`bspc subscribe -f` makes bspwm freeze.

### ISSUE THREAD

**Issue by emanuele6** [baskerville/bspwm#1305]
A reddit user discovered that running `bspc subscribe -f` makes bspwm freeze ([reddit post on r/bspwm](https://www.reddit.com/r/bspwm/comments/o2255q/how_to_make_smartdynamic_gap_sizes_what_is_gap/)).

I investigated this issue running `bspwm` with `gdb` adding a breakpoint at `src/messages.c:1332`.

It seems that the `fopen()` at `src/messages.c:1355` locks.
https://github.com/baskerville/bspwm/blob/25f2cd17011d4032d255de41f70581b87c90cbcc/src/messages.c#L1355

I found [this article on ibm.com](https://www.ibm.com/support/pages/apar/PK07789) that states:
> If the fopen() function is used to open named pipes in a multi-threaded environment, a deadlock will occur.  The deadlock is caused by a named pipe waiting for the other end of the pipe to be opened, while still holding the fopen() multi-thread mutex.

I'm not sure if that is the problem though.

**Comment by emanuele6** [comment]
opening `fifo_path` with `"r+"` instead of `"w"` doesn't cause a deadlock.

```diff
diff --git a/src/messages.c b/src/messages.c
index bca55c9..5640016 100644
--- a/src/messages.c
+++ b/src/messages.c
@@ -1352,7 +1352,7 @@ void cmd_subscribe(char **args, int num, FILE *rsp)
 		fflush(rsp);
 		fclose(rsp);
 
-		stream = fopen(fifo_path, "w");
+		stream = fopen(fifo_path, "r+");
 
 		if (stream == NULL) {
 			perror("subscribe: fopen");
```

**Comment by ortango** [comment]
i thought this was common knowledge, but it would be neat if it didn't block without a reader.

but, is `O_RDWR` *on a fifo* ok to use outside of linux?

openbsd's `open(2)` manual says not to use it.

>A FIFO should either be opened with O_RDONLY or with O_WRONLY. The behavior for opening a FIFO with O_RDWR is undefined.

**Comment by emanuele6** [comment]
> i thought this was common knowledge, but it would be neat if it didn't block without a reader.

Oh, yesterday, I had tried reading the fifo after running `bspc subscribe -f`, but i remember it still locked when I tried, that is why I investigated further.
I thought `bspc subscribe -f` didn't work at all (I never used it).

I probably did that test wrong; I see now that that `fopen("w")` returns and let's bspwm continue when some process reads from the fifo.

 This probably means that the issue is that it is only possible to open named pipe for writing after they are also opened for reading by another process.
 
 I guess this issue can be closed since this is not a bug and `bspc subscribe -f` works correctly.
